Public/Src/Pips/Dll/Operations/Pip.cs

```diff
@@ -388,5 +389,18 @@ public string GetShortDescription(PipExecutionContext context)
 
             return $"{moduleName} - {valueName}{toolName} [{qualifierName}]";
         }
+
+        /// <summary>
+        /// Check if there is any output of this pip are affected by the source change
+        /// </summary>
+        public virtual bool IsOutputAffectedBySourceChange(

```

Comment:
>IsOutputAffectedBySourceChange( [](start = 28, length = 31)

The pattern that I see in the code is pips are treated as data. Yes, your approach is more object oriented, but given that pips are data, adding a method here may not be proper. 

All the information you need for pips' inputs/outputs can be retrieved using our `PipArtifacts.cs`. Can you use that class instead?